databricks/koalas/groupby.py

```diff
@@ -602,6 +602,48 @@ def rename_output(pdf):
             sdf=sdf, data_columns=return_schema.fieldNames(), index_map=[])  # index is lost.
         return DataFrame(internal)
 
+    def nunique(self, dropna=True):
+        """
+        Return DataFrame with number of distinct observations per group for each column.
+
+        Parameters
+        ----------
+        dropna : boolean, default True
+            Don’t include NaN in the counts.
+
+        Returns
+        -------
+        nunique : DataFrame
+
+        Examples
+        --------
+
+        >>> df = ks.DataFrame({'id': ['spam', 'egg', 'egg', 'spam',
+        ...                           'ham', 'ham'],
+        ...                    'value1': [1, 5, 5, 2, 5, 5],
+        ...                    'value2': list('abbaxy')})
+        >>> df
+             id  value1 value2
+        0  spam       1      a
+        1   egg       5      b
+        2   egg       5      b
+        3  spam       2      a
+        4   ham       5      x
+        5   ham       5      y
+
+        >>> df.groupby('id').nunique() # doctest: +NORMALIZE_WHITESPACE
+              id  value1  value2
+        id
+        egg    1       1       1
+        ham    1       1       2
+        spam   1       2       1
+        """
+        if not dropna:
+            raise ValueError('nunique do not support `drop=False` now')
+        if not self._have_agg_columns:
+            self._agg_columns = self._groupkeys + self._agg_columns
+        return self._reduce_for_stat_function(F.countDistinct, only_numeric=False)
```

Comment:
There's `DataFrame.nunique` implementation. Can we follow and ideally deduplicate the logic?

Reply:
Can you give me some more specific code?

Reply:
You can make a private method called, for instance, `_ nunique` at `DataFrame`.

```python
def _nunique(c, dropna=True):
    count_fn = partial(F.approx_count_distinct, rsd=rsd) if approx else F.countDistinct
    if dropna:
        return count_fn(c).alias(c)
    else:
        return count_fn(c) + F.when(F.count(F.when(self._internal.scol_for(c).isNull(), 1)
                                                  .otherwise(None))
                                          >= 1, 1).otherwise(0))
```

and reuse this here and `DataFrame. nunique`

Reply:
```python
    def _nunique(self, col, dropna=True, approx=False, rsd=0.05):
        count_fn = partial(F.approx_count_distinct, rsd=rsd) if approx else F.countDistinct
        if dropna:
            return count_fn(col).alias(col)
        else:
            return (count_fn(col) +
                    F.when(F.count(F.when(self._internal.scol_for(col).isNull(), 1)
                                   .otherwise(None)) >= 1, 1).otherwise(0)).alias(col)
```
Excuse me, i add the `_nunique` function in `DataFrame`.But i want to use it in `groupby`  by using `_reduce_for_stat_function`. It raise errors `_nunique() missing 1 required positional argument: 'col'`.
How can i send `col` to `_nunique`? 
```
        def nunique(self, dropna=True):
        if isinstance(self, DataFrameGroupBy):
            self._agg_columns = self._groupkeys + self._agg_columns
        return self._reduce_for_stat_function(self._kdf._nunique, only_numeric=False)
```

Reply:
Okay, let me maybe try it later. Ignore this comment for now.

Reply:
emmm........ Can you cc this?